rfcs/20221006-frontend_event_tracking.md

```diff
@@ -0,0 +1,296 @@
+- The ideal data warehouse for our frontend events
+- Whether to use a SaaS product or self-host analytics software
+
+I suggest that our frontend events should be agnostic to any specific analytics platform. This means that our frontend events will be generalized to contain information we think is useful, and munged as necessary to conform to the needs of our analytics service. I see many benefits to doing this:
+
+- It avoids vendor lock-in to a specific analytics platform
+- It makes it much easier for users to audit exactly what values we're collecting, by avoiding heavy and mysterious 3rd-party SDKs
+- Makes us less likely to be blocked by ad and script blocking software
+
+There are many established conventions around frontend event collection. Using our own system should allow us to adopt most 3rd party platforms without locking us in. A quick example use case is the Google Analytics alternative [Plausible.io](https://plausible.io/), which includes an [events API](https://plausible.io/docs/events-api). Plausible accepts the following simple request shape to capture a pageview, for example:
+
+```shell
+curl -i -X POST https://plausible.io/api/event \
+  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36 OPR/71.0.3770.284' \
+  -H 'X-Forwarded-For: 127.0.0.1' \
+  -H 'Content-Type: application/json' \
+  --data '{"name":"pageview","url":"http://dummy.site","domain":"dummy.site","screen_width":1666}'
+```
+
+There are many points in the code execution at which we could transform our custom events into ones supported by other platforms. By adopting and implementing our own custom solution, we can start collecting data as soon as possible without being blocked on time spent evaluating and implementing a full analytics system.
+
+## The technical details
+
+We will create a system for collecting events on the frontend. Each event will have a unique name, collect some default data about the enviroment (timestamp, referer, and so on) and can have an optional data payload containing string and numerical values. I'd like to limit the payload shape to strings and numbers to keep it JSON serializable and simple. Allowing for complex nested objects might prove too complex to query in the future. I think this is a reasonable constraint but am open to alternatives.
+
+### Sending events
+
+These events will be sent as `POST` requests to a server route on the frontend, which will then route them to our desired analytics service. In the future, it may make sense to move this server route to a standalone microservice, but for now a [Nuxt server route](https://v3.nuxtjs.org/guide/directory-structure/server/#server-routes) should be sufficient to handle the analytics events. This endpoint is essentially a proxy between the frontend and our 'true' analytics service that makes sure we're always sending events server-side.
```

Comment:
Since we are still on Nuxt 2 😢, we should probably also include [Nuxt serverMiddleware](https://nuxtjs.org/docs/configuration-glossary/configuration-servermiddleware/) as a Nuxt 2 alternative to Nuxt server routes.

Reply:
Mentioned in 7265661.